UX improvements for `tctl` with Admin MFA

Joerger opened the pull request:
Changelog: UX improvements for `tctl` with Admin MFA

Changes:
- Prompt once for bulk `tctl create` requests
- Prompt once for `tctl edit` before opening the editor

Depends on https://github.com/gravitational/teleport/pull/51618

Closes https://github.com/gravitational/teleport/issues/38530
Closes https://github.com/gravitational/teleport/issues/38532

Joerger committed:
Prompt admin mfa once for creating multiple resources.

Joerger committed:
Disable bulk create test for auth preference.

Joerger committed:
Prompt admin mfa before tctl edit.

rosstimothy reviewed (changes requested):
Hrm this actually looks like it breaks tctl create -f with multiple resources.

```bash
$ cat shu2.yaml 
---
kind: static_host_user
version: v2
metadata:
  name: george
spec:
  matchers:
    - node_labels:
      - name: "foo"
        values: ["bar"]
      groups:
        - foo
        - bar
      uid: 54321
      gid: 1038 
    - node_labels_expression: 'labels["foo"] == labels["bar"]'
      groups:
        - baz
        - quux
      uid: 324
      gid: 678
      sudoers: ["ALL = (root) NOPASSWD: /usr/bin/systemctl restart nginx.service"]
---
kind: static_host_user
version: v2
metadata:
  name: george2
spec:
  matchers:
    - node_labels:
      - name: "foo"
        values: ["bar"]
      groups:
        - foo
        - bar
      uid: 54321
      gid: 1038
    - node_labels_expression: 'labels["foo"] == labels["bar"]'
      groups:
        - baz
        - quux
      uid: 324
      gid: 678
      sudoers: ["ALL = (root) NOPASSWD: /usr/bin/systemctl restart nginx.service"]


$ build/tctl create -f resources/shu2.yaml
This is an admin-level action and requires MFA to complete
Tap any security key
Detected security key tap
ERROR: admin-level API request requires MFA verification
```

It looks like tctl has been updated to only prompt once and reuse the token, however not all of the corresponding Create/Upsert RPCs have been converted to use `AuthorizeAdminActionAllowReusedMFA` instead of `AuthorizeAdminAction`.

Joerger commented on lib/authz/permissions.go:
```diff
@@ -585,7 +585,15 @@ func (a *authorizer) authorizeAdminAction(ctx context.Context, authContext *Cont
+		// invalid MFA verification should be a noop, this way an actionable error will be
+		// returned instead. e.g. "admin-level API request requires MFA verification".
+		//
+		// TODO(Joerger): Currently we get a validation error when a reusable Webauthn response
+		// is provided but it's session data has been overwritten. This happens in bulk
+		// `tctl create` requests with endpoints mixed in that disallow reuse. Ignoring the error
+		// here is harmless, but a better way to handle this would be to add a separate index for
+		// reusable Webauthn session data.
```
@codingllama Any thoughts on this? Right now we just have `login` and `register`. I think we'd just need to add `login_with_reuse`. When applicable, we would just check for `login`, `login_with_reuse`, or both. I could do this in a follow up PR if it sounds reasonable to you.

codingllama replied:
I'm not sure I follow the suggestion. Isn't this what allow reuse is for? If we want to allow challenge reuse for some resources, shouldn't we do that instead?

Joerger replied:
My concern is that even if we make all `tctl create` endpoints allow reuse now, we might add one in the future which doesn't, which would cause the flow to breakdown completely with no way around it (due to the issue explained here). I suppose we could just be careful not to screw it up and worst case do a patch fix if it happens.

Joerger commented:
@rosstimothy Fixed the issue you reported in 6f1dfa170ecab83705a964e8485a524946332f68. Note that having a resource which disallows reuse in the middle of the flow causes it to prompt a second time, plus another time for every resource that follows. Meaning worst case you'd have n+1 prompt. See the TODO on how we could fix this in the webauthn backend.

We can also probably update more endpoints to allow reuse but we should do this on a case by case basis and support non-reusable MFA in this flow anyways.

codingllama commented on api/utils/grpc/interceptors/mfa.go:
```diff
@@ -44,8 +45,16 @@ func WithMFAUnaryInterceptor(mfaCeremony mfa.CeremonyFn) grpc.UnaryClientInterce
+		} else if mfaResp != nil && mfaResp.GetWebauthn() != nil {
+			// TODO(Joerger): DELETE IN v18.0.0
+			// If we passed an WebAuthn response and got an error, it may be that the
+			// backend WebAuthn session data got overwritten. Fallback to retrying with
```
This is a strong assumption to make, we shouldn't be guessing. We are also assuming WebAuthn only ever fails this way, which isn't true.

Same for other places where this happens.

codingllama commented on lib/auth/auth_with_roles.go:
```diff
@@ -5755,15 +5759,15 @@ func (a *ServerWithRoles) IsMFARequired(ctx context.Context, req *proto.IsMFAReq
 	// Check if MFA is required for admin actions. We don't currently have
 	// a reason to check the name of the admin action in question.
 	if _, ok := req.Target.(*proto.IsMFARequiredRequest_AdminAction); ok {
-		if a.context.AdminActionAuthState == authz.AdminActionAuthUnauthorized {
```
Could you give some background on this change?

Joerger replied:
I split it off into a separate PR - https://github.com/gravitational/teleport/pull/51618

codingllama commented on lib/authz/permissions.go:
```diff
@@ -585,7 +585,15 @@ func (a *authorizer) authorizeAdminAction(ctx context.Context, authContext *Cont
+		// `tctl create` requests with endpoints mixed in that disallow reuse. Ignoring the error
+		// here is harmless, but a better way to handle this would be to add a separate index for
+		// reusable Webauthn session data.
+		return nil
```
Swallowing errors from ValidateMFAAuthResponse looks wrong to me. MFA/WebAuthn verification can fail in a variety of ways, many of them legitimate.

Joerger commented:
@codingllama @rosstimothy This is ready for another look.

https://github.com/gravitational/teleport/pull/49095 unblocked this PR by enabling reuse for all of the endpoints hit by tctl create -f multiple-resources.yaml. With that I was able to remove all the questionable logic (switching between reusable/non-reusable mfa mid request).

codingllama commented:
Are the test failures genuine?

Joerger commented:
> Are the test failures genuine?

Well I forgot to push, so yes 😅

Joerger committed:
Get reusable MFA for listing tokens or cert authorities.

Joerger committed:
Don't prompt for mfa twice when editing cert authority.

Joerger commented:
@codingllama I fixed the issue with the test and added additional test cases for `tctl get cert_authorities` and `tctl get tokens`, if you want to take another look.

@rosstimothy This is ready for re-review

codingllama commented on lib/auth/auth_with_roles.go:
```diff
@@ -5984,15 +5984,15 @@ func (a *ServerWithRoles) IsMFARequired(ctx context.Context, req *proto.IsMFAReq
 	// a reason to check the name of the admin action in question.
 	if _, ok := req.Target.(*proto.IsMFARequiredRequest_AdminAction); ok {
-		if a.context.AdminActionAuthState == authz.AdminActionAuthUnauthorized {
+		if a.context.AdminActionAuthState == authz.AdminActionAuthNotRequired {
```
Isn't this part of #51618? Missing a rebase?

codingllama commented on tool/tctl/common/admin_action_test.go:
```diff
@@ -384,6 +387,14 @@ func (s *adminActionTestSuite) testTokens(t *testing.T) {
 	}
 
+	createTokens := func() error {
+		return trace.NewAggregate(s.authServer.CreateToken(ctx, token), s.authServer.CreateToken(ctx, token2))
```
Split lines to the 2 operations are more apparent?

```suggestion
		return trace.NewAggregate(
			s.authServer.CreateToken(ctx, token),
			s.authServer.CreateToken(ctx, token2),
		)
```

codingllama commented on tool/tctl/common/admin_action_test.go:
```diff
@@ -384,6 +387,14 @@ func (s *adminActionTestSuite) testTokens(t *testing.T) {
+	}
+
+	deleteTokens := func() error {
+		return trace.NewAggregate(s.authServer.DeleteToken(ctx, token.GetName()), s.authServer.DeleteToken(ctx, token2.GetName()))
```
```suggestion
		return trace.NewAggregate(
			s.authServer.DeleteToken(ctx, token.GetName()),
			s.authServer.DeleteToken(ctx, token2.GetName()),
		)
```

codingllama commented on tool/tctl/common/admin_action_test.go:
```diff
@@ -488,11 +517,21 @@ func (s *adminActionTestSuite) testCertAuthority(t *testing.T) {
 		return s.authServer.DeleteCertAuthority(ctx, ca.GetID())
 	}
 
+	createCertAuthorities := func() error {
+		return trace.NewAggregate(s.authServer.CreateCertAuthority(ctx, ca), s.authServer.CreateCertAuthority(ctx, ca2))
+	}
+
+	deleteCertAuthorities := func() error {
+		return trace.NewAggregate(s.authServer.DeleteCertAuthority(ctx, ca.GetID()), s.authServer.DeleteCertAuthority(ctx, ca2.GetID()))
+	}
```
```suggestion
	createCertAuthorities := func() error {
		return trace.NewAggregate(
			s.authServer.CreateCertAuthority(ctx, ca),
			s.authServer.CreateCertAuthority(ctx, ca2),
		)
	}

	deleteCertAuthorities := func() error {
		return trace.NewAggregate(
			s.authServer.DeleteCertAuthority(ctx, ca.GetID()),
			s.authServer.DeleteCertAuthority(ctx, ca2.GetID()),
		)
	}
```

Joerger committed:
Address comment.

Joerger commented on tool/tctl/common/edit_command.go:
```diff
@@ -127,6 +128,14 @@ func (e *EditCommand) editResource(ctx context.Context, client *authclient.Clien
 	}
 	rc.Initialize(e.app, nil, e.config)
 
+	// Prompt for admin action MFA if required, before getting any
+	// resources with secrets and before the update/editor below.
+	if mfaResponse, err := mfa.PerformAdminActionMFACeremony(ctx, client.PerformMFACeremony, true /*allowReuse*/); err == nil {
+		ctx = mfa.ContextWithMFAResponse(ctx, mfaResponse)
+	} else if !errors.Is(err, &mfa.ErrMFANotRequired) && !errors.Is(err, &mfa.ErrMFANotSupported) {
+		return trace.Wrap(err)
+	}
```
The difference is that you don't lose your edits when failing/cancelling the MFA attempt in the WebUI. In `tctl` with the terminal editor, it's cumbersome to have to go back and repeat your edits from memory.

Joerger commented on tool/tctl/common/resource_command.go:
```diff
@@ -288,6 +288,20 @@ func (rc *ResourceCommand) GetRef() services.Ref {
 
 // Get prints one or many resources of a certain type
 func (rc *ResourceCommand) Get(ctx context.Context, client *authclient.Client) error {
+	// If at least one ref is for a cert authority with secrets, prompt for admin mfa.
+	if _, err := mfa.MFAResponseFromContext(ctx); trace.IsNotFound(err) {
+		if rc.withSecrets && slices.ContainsFunc(rc.refs, func(r services.Ref) bool {
+			return r.Kind == types.KindCertAuthority
+		}) {
+			mfaResponse, err := mfa.PerformAdminActionMFACeremony(ctx, client.PerformMFACeremony, true /*allowReuse*/)
+			if err == nil {
+				ctx = mfa.ContextWithMFAResponse(ctx, mfaResponse)
+			} else if !errors.Is(err, &mfa.ErrMFANotRequired) && !errors.Is(err, &mfa.ErrMFANotSupported) {
+				return trace.Wrap(err)
+			}
+		}
+	}
+
```
Perhaps, but this error is being returned by the backend on attempting `ListUsers`. I'm not sure that we should update `tctl` to assume this backend logic. WDYT?

This is the existing behavior before this PR, since we successfully perform other admin actions within `tctl get all` before failing on `ListUsers`. Before this PR, it can actually prompt zero to multiple times depending on how go decides to order the `map[string]ResourceMarshaler` keys (resource `Kind`):

```
> tctl get all
ERROR: this request can be only executed by an admin
> tctl get all
This is an admin-level action and requires MFA to complete
Tap any security key
Detected security key tap
This is an admin-level action and requires MFA to complete
Tap any security key
Detected security key tap
ERROR: this request can be only executed by an admin
```

Making it deterministically prompt once is definitely an improvement.

Joerger committed:
Fix getting multiple tokens; add test for getting multiple, not all, tokens or cert authorities; clean up mfa logic for get.

Joerger commented:
@codingllama @rosstimothy I did a final passover and noticed that getting multiple tokens still prompted MFA multiple times. I fixed this and added a unit test for `GetMany`, e.g. `tctl get token/1,token/2` - d956be6028cb311c2071714f5106e30960fc670c

Joerger committed:
Merge branch 'master' into joerger/tctl-create-prompt-once
